core/src/main/java/com/linecorp/armeria/common/DefaultHttpResponseBodyStream.java

```diff
@@ -0,0 +1,352 @@
+import com.google.common.math.LongMath;
+
+import com.linecorp.armeria.common.stream.NoopSubscriber;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class DefaultHttpResponseBodyStream implements HttpResponseBodyStream {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultHttpResponseBodyStream.class);
+
+    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscription.class, "upstream");
+
+    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscriber> downstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscriber.class, "downstream");
+
+    private static final ResponseHeaders HEADERS_WITH_UNKNOWN_STATUS = ResponseHeaders.of(HttpStatus.UNKNOWN);
+    private static final HeadersFuture<List<ResponseHeaders>> EMPTY_INFORMATIONAL_HEADERS;
+    private static final HeadersFuture<HttpHeaders> EMPTY_TRAILERS;
+
+    static {
+        EMPTY_INFORMATIONAL_HEADERS = new HeadersFuture<>();
+        EMPTY_INFORMATIONAL_HEADERS.doComplete(ImmutableList.of());
+
+        EMPTY_TRAILERS = new HeadersFuture<>();
+        EMPTY_TRAILERS.doComplete(HttpHeaders.of());
+    }
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<DefaultHttpResponseBodyStream, HeadersFuture>
+            informationalHeadersFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultHttpResponseBodyStream.class, HeadersFuture.class, "informationalHeadersFuture");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<DefaultHttpResponseBodyStream, HeadersFuture>
+            trailersFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultHttpResponseBodyStream.class, HeadersFuture.class, "trailersFuture");
+
+    private final HeadersFuture<ResponseHeaders> headersFuture = new HeadersFuture<>();
+    private final BodySubscriber bodySubscriber = new BodySubscriber();
+    private final HttpResponse response;
+    private final EventExecutor executor;
+
+    @Nullable
+    private volatile HeadersFuture<List<ResponseHeaders>> informationalHeadersFuture;
+    @Nullable
+    private volatile HeadersFuture<HttpHeaders> trailersFuture;
+    private volatile boolean wroteAny;
+
+    DefaultHttpResponseBodyStream(HttpResponse response, EventExecutor executor,
+                                  SubscriptionOption... options) {
+        requireNonNull(response, "response");
+        this.response = response;
+        this.executor = executor;
+
+        response.subscribe(bodySubscriber, executor, options);
+        // Prefetch headers
+        bodySubscriber.request(1);
```

Comment:
How about doing this is the `bodySubscriber`? (e.g. setting `pendingRequests` to 1, etc.)